From cc5e793eaf9e6a52daffd08c22c808cc3c651f68 Mon Sep 17 00:00:00 2001
From: Albert Astals Cid <aacid@kde.org>
Date: Thu, 9 Oct 2025 10:11:13 +0200
Subject: [PATCH 396/553] appendNodeText: Don't split emojis with surrogate
 chars in two

When parsing a document that assigns an anchor to text, we assign
this anchor to the next immediate character. If this character was the
low surrogate in a pair, we would end up splitting the surrogate and
messing up the text. We use QStringIterator to make sure we always
process surrogate pairs together and add the full pair to the resulting
text.

Fixes: QTBUG-140929
Change-Id: I63ee03c251004d1a138f97462723fcc3798b9147
Reviewed-by: Albert Astals Cid <aacid@kde.org>
Reviewed-by: Eskil Abrahamsen Blomfeldt <eskil.abrahamsen-blomfeldt@qt.io>
(cherry picked from commit f9b83106255aa97d0a3ffc105fb17676d2a72b78)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/gui/text/qtextdocumentfragment.cpp | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/src/gui/text/qtextdocumentfragment.cpp b/src/gui/text/qtextdocumentfragment.cpp
index 1b6e76c2017..5797d1a68b4 100644
--- a/src/gui/text/qtextdocumentfragment.cpp
+++ b/src/gui/text/qtextdocumentfragment.cpp
@@ -16,6 +16,7 @@
 #include <qbytearray.h>
 #include <qdatastream.h>
 #include <qdatetime.h>
+#include <QtCore/private/qstringiterator_p.h>
 
 QT_BEGIN_NAMESPACE
 
@@ -582,8 +583,11 @@ bool QTextHtmlImporter::appendNodeText()
     QString textToInsert;
     textToInsert.reserve(text.size());
 
-    for (QChar ch : text) {
-        if (ch.isSpace()
+    QStringIterator it(text);
+    while (it.hasNext()) {
+        char32_t ch = it.next();
+
+        if (QChar::isSpace(ch)
             && ch != QChar::Nbsp
             && ch != QChar::ParagraphSeparator) {
 
@@ -646,12 +650,12 @@ bool QTextHtmlImporter::appendNodeText()
 
                 format.setAnchor(true);
                 format.setAnchorNames(namedAnchors);
-                cursor.insertText(ch, format);
+                cursor.insertText(QString::fromUcs4(&ch, 1), format);
                 namedAnchors.clear();
                 format.clearProperty(QTextFormat::IsAnchor);
                 format.clearProperty(QTextFormat::AnchorName);
             } else {
-                textToInsert += ch;
+                textToInsert += QChar::fromUcs4(ch);
             }
         }
     }
-- 
2.51.2

